centrifuge_cli/main.py

```
@@ -278,39 +305,84 @@ def code_emulated(cli, exid, path):
     if exid and path:
         query_list.append(f'path={path}')
 
-    click.echo(cli.do_GET(f'/api/report/{cli.ufid}/emulated-files/{exid}', query_list=query_list))
+    result = cli.do_GET(f'/api/report/{cli.ufid}/emulated-files/{exid}', query_list=query_list)
+    cli.echo(result)
+    return(result)
 
 
 @report.command()
 @pass_cli
 def certificates(cli):
-    click.echo(cli.do_GET(f'/api/report/crypto/{cli.ufid}/certificates', get_all=True))
+    result = cli.do_GET(f'/api/report/crypto/{cli.ufid}/certificates', get_all=True)
+    cli.echo(result)
+    return(result)
 
 
 @report.command(name='private-keys')
 @pass_cli
-def privatekeys(cli):
-    click.echo(cli.do_GET(f'/api/report/crypto/{cli.ufid}/privateKeys', get_all=True))
+def private_keys(cli):
+    result = cli.do_GET(f'/api/report/crypto/{cli.ufid}/privateKeys', get_all=True)
+    cli.echo(result)
+    return(result)
 
 
 @report.command(name='public-keys')
 @pass_cli
 def public_keys(cli):
-    click.echo(cli.do_GET(f'/api/report/crypto/{cli.ufid}/publicKeys', get_all=True))
+    result = cli.do_GET(f'/api/report/crypto/{cli.ufid}/publicKeys', get_all=True)
+    cli.echo(result)
+    return(result)
 
 
 @report.command(name='binary-hardening')
 @pass_cli
 def binary_hardening(cli):
-    click.echo(cli.do_GET(f'/api/report/{cli.ufid}/binary-hardening', get_all=True))
+    result = cli.do_GET(f'/api/report/{cli.ufid}/binary-hardening', get_all=True)
+    cli.echo(result)
+    return(result)
+
+
+@report.command(name='check-policy')
+@click.option('--policy-yaml', metavar='FILE', type=click.Path(), help='Centrifuge policy yaml file.', required=True)
+@click.pass_context
+@pass_cli
+def check_policy(cli, ctx, policy_yaml):
+    outfmt = cli.outfmt
+    cli.outfmt = 'json'
+    cli.echo_enabled = False
+    cli.limit = 1000
+    certificates_json = json.loads(ctx.invoke(certificates))
```

Centrifuge CLI methods have been modified to throw errors such as Authorization and Connection related errors. Those are simple user-friendly messages in string format so directly converting to JSON here will fail. I believe we have to add a pre-check here to check if Authtoken and centrifuge server is valid. If not then probably we error out the CLI script.

---

Good point. Instead of turning those exceptions into json error messages we should just throw an exception or not catch it so it bubbles back up to the user's console. 

only reason those were added is so you could parse the error when using the cli as a subprocess. the exception can be thrown and printed to the console.

---

I disagree. Users that use the CLI would likely rather see the stack trace than a useless error message that might supress the actual error.

---

ok that's a fair point.  if we are suppressing the real error that's no good.  i'm just going by the 2 users that were trying to use this and sent in requests for help because it wasn't working for them and was spitting out stack traces similar to this one:

```
Traceback (most recent call last):

  File "/usr/local/bin/centrifuge", line 10, in <module>

    sys.exit(cli())

  File "/usr/lib/python3/dist-packages/click/core.py", line 764, in __call__

    return self.main(*args, **kwargs)

  File "/usr/lib/python3/dist-packages/click/core.py", line 717, in main

    rv = self.invoke(ctx)

  File "/usr/lib/python3/dist-packages/click/core.py", line 1137, in invoke

    return _process_result(sub_ctx.command.invoke(sub_ctx))

  File "/usr/lib/python3/dist-packages/click/core.py", line 1137, in invoke

    return _process_result(sub_ctx.command.invoke(sub_ctx))

  File "/usr/lib/python3/dist-packages/click/core.py", line 956, in invoke

    return ctx.invoke(self.callback, **ctx.params)

  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke

    return callback(*args, **kwargs)

  File "/usr/lib/python3/dist-packages/click/decorators.py", line 64, in new_func

    return ctx.invoke(f, obj, *args, **kwargs)

  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke

    return callback(*args, **kwargs)

  File "/usr/local/lib/python3.7/dist-packages/centrifuge_cli/main.py", line 257, in certificates

    click.echo(cli.do_GET(f'/api/report/crypto/{cli.ufid}/certificates', get_all=True))

  File "/usr/local/lib/python3.7/dist-packages/centrifuge_cli/main.py", line 67, in do_GET

    res.raise_for_status()

  File "/usr/lib/python3/dist-packages/requests/models.py", line 940, in raise_for_status

    raise HTTPError(http_error_msg, response=self)

requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://centrifuge.refirmlabs.com/api/report/crypto/***/certificates?limit=1000&authtoken=******&page=1
```

The problem was that they weren't talking to the right centrifuge server, but they weren't able to determine that from the stack  trace.  They might be outliers as far as user background knowledge, but so far they make up 100% of the beta test users.  Open to better ways to let them know that the script failed due to them not supplying the correct combination of hostname and authentication token.

---

@rflmk I can see the argument for 403 response to print something friendly since that is a common error condition. 

However for other unforeseen errors, would just showing the last line of that stack trace help the user self-resolve the issue better? If we suppress the stack trace altogether and just show the message I don't believe the outcome would be any different, users will still put in a support request.

Happy to suppress the stack trace and just display the message portion but I don't think that fixes the issue.

---

Btw `raise SystemExit` does what I think you want. No more stack traces just the error message then exit. So I think we arrived at what you wanted by accident :)